spcs-three-tier/app/setup.sql

```diff
@@ -47,26 +47,33 @@ BEGIN
 END;	
 $$;
 
-GRANT USAGE ON PROCEDURE v1.get_configuration(STRING) TO APPLICATION ROLE app_admin;
+GRANT USAGE ON PROCEDURE versioned_schema.get_configuration(STRING) TO APPLICATION ROLE app_admin;
 
 -- The version initializer callback is executed after a successful installation, upgrade, or downgrade of an application object.
--- In case the application fails to upgrade, the version initializer of the previous (successful) version will be executed so you 
+-- In case the application fails to upgrade, the version initializer of the previous (successful) version will be executed so you
 -- can clean up application state that may have been modified during the failed upgrade.
-CREATE OR REPLACE PROCEDURE v1.init()
-RETURNS STRING 
+
+-- this is the first the version(version v1 patch 0) of the app package. We consider the case that when the
+-- app is upgraded to next version and try to alter the services and it fails. In that case
+-- it will fail back to version v1 patch 0 and call this procedure versioned_schema.init() to
+-- restore the services and app can fully function.
+CREATE OR REPLACE PROCEDURE versioned_schema.init()
+RETURNS STRING
 LANGUAGE SQL
-EXECUTE AS OWNER 
+EXECUTE AS OWNER
 AS
 $$
-BEGIN    
-    ALTER SERVICE IF EXISTS app_public.frontend FROM SPECIFICATION_FILE='frontend.yaml';
-    ALTER SERVICE IF EXISTS app_public.backend FROM SPECIFICATION_FILE='backend.yaml';
+BEGIN
+    --create services if not exist
+    call versioned_schema.create_services();
+    --wait for 300 seconds, but it will stop if two services have READY status,
+    -- or any of the services has the FAILED status.
+    select system$wait_for_services(300, 'app_public.backend', 'app_public.frontend');
+
```

Comment:
Handle the failure to  create service
Should throw exception  to force app failure?

Reply:
Do you know what the best practice is from the NA+SPCS team? That's what we want to have here.

Reply:
Had a test and found out that if the service was not available or not ready, it will fail